openstack_controller/datadog_checks/openstack_controller/api.py

```diff
@@ -0,0 +1,222 @@
+# (C) Datadog, Inc. 2018
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import requests
+import simplejson as json
+from six.moves.urllib.parse import urljoin
+from .settings import DEFAULT_API_REQUEST_TIMEOUT, DEFAULT_KEYSTONE_API_VERSION, DEFAULT_NEUTRON_API_VERSION
+from .exceptions import (InstancePowerOffFailure, AuthenticationNeeded, KeystoneUnreachable)
+
+
+class AbstractApi:
+    DEFAULT_API_REQUEST_TIMEOUT = 10  # seconds
+
+    def __init__(self, logger, ssl_verify, proxy_config):
+        self.logger = logger
+        self.ssl_verify = ssl_verify
+        self.proxy_config = proxy_config
+
+    def _make_request(self, url, headers=None, params=None, timeout=DEFAULT_API_REQUEST_TIMEOUT):
+        """
+        Generic request handler for OpenStack API requests
+        Raises specialized Exceptions for commonly encountered error codes
+        """
+        self.logger.debug("Request URL and Params: %s, %s", url, params)
+        resp = {}
+        try:
+            resp = requests.get(
+                url,
+                headers=headers,
+                verify=self.ssl_verify,
+                params=params,
+                timeout=timeout,
+                proxies=self.proxy_config,
+            )
+            resp.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            self.logger.debug("Error contacting openstack endpoint: %s", e)
+            if resp.status_code == 401:
+                self.logger.info('Need to reauthenticate before next check')
+                raise AuthenticationNeeded()
+            elif resp.status_code == 409:
+                raise InstancePowerOffFailure()
+            else:
+                raise e
+
+        return resp.json()
+
+
+class ComputeApi(AbstractApi):
+    def __init__(self, logger, ssl_verify, proxy_config, endpoint, auth_token):
+        AbstractApi.__init__(self, logger, ssl_verify, proxy_config)
+        self.endpoint = endpoint
+        self.auth_token = auth_token
+
+    def get_os_hypervisor_uptime(self, hyp_id):
+        url = '{}/os-hypervisors/{}/uptime'.format(self.endpoint, hyp_id)
+        headers = {'X-Auth-Token': self.auth_token}
+        resp = self._make_request(url, headers)
+        return resp['hypervisor']['uptime']
```

Comment:
We may want to try/catch this in case the endpoint returns some unexpected result that doesn't include `uptime`